### PR TITLE
HPCC-20133 Upgrade ESDL transaction log for Royalty logging, etc

### PR DESCRIPTION
**Pull request by wangkx**
1. Pass royalty dataset to log content in order to support
Royalty logging;
2. the transaction ID created by the existing code always has '-'
or '-X' between transaction seed and sequence number. Now, the
characters are configurable;
3. Create new class CLogContentFilter. Move the content of
CESPServerLoggingAgent.filterLogContent() into CLogContentFilter
.filterLogContent() so that it can be shared with other log
agents, such as WsLogServiceESPAgent. Also move the following
methods from CESPServerLoggingAgent to CLogContentFilter:
readAllLogFilters(), readLogFilters, addLogContentBranch,
filterAndAddLogContentBranch, and filterLogContentTree;
4. Move the following items from espserviceloggingagent.hpp to
loggingagentbase.hpp so that they can be shared for other log
agents: enum ESPLogContentGroup, const espLogContentGroupNames[],
class CTransIDBuilder, and class CESPLogContentGroupFilters.
5. Move wslogserviceespagent.xsl from LN repo to here because it
is needed by esp_service_DynamicESDL.xsl. Add code into
esp_service_DynamicESDL.xsl to call the template in
wslogserviceespagent.xsl
6. Move the xsl code of the transactionId/seed from logging_agent.xsl
to new esp_log_transid.xsl so that it can be shared by other log
agents.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review this PR and https://github.com/hpcc-systems/LN/pull/891.

**Review by afishbeck — COMMENTED** [review]
@wangkx looks ok to me.
@gaojx who can review this from the scapps side?

File: esp/logging/logginglib/loggingagentbase.cpp [hunk 1]
```diff
@@ -22,6 +22,285 @@ static const char* const defaultTransactionTable = "transactions";
...
+void CLogContentFilter::addLogContentBranch(StringArray& branchNames, IPropertyTree* contentToLogBranch, IPropertyTree* updateLogRequestTree)
+{
+    IPropertyTree* pTree = updateLogRequestTree;
+    unsigned numOfBranchNames = branchNames.length();
+    if (numOfBranchNames > 0)
+    {
+        unsigned i = 0;
+        while (i < numOfBranchNames)
+        {
+            const char* branchName = branchNames.item(i);
```
Comment (gaojx) [inline esp/logging/logginglib/loggingagentbase.cpp]: The outer "if" is not needed. Here a "for" loop is clearer than a "while" loop.

**Comment by wangkx** [comment]
@jiafu I made the changes based on your comment. Please review my changes.

**Review by gaojx — APPROVED** [review]

File: esp/logging/loggingagent/espserverloggingagent/loggingagent.cpp [hunk 2]
```diff
@@ -31,6 +31,10 @@ static const char* const PropServerWaitingSeconds = "MaxServerWaitingSeconds";
 static const char* const PropMaxTransIDLength = "MaxTransIDLength";
 static const char* const PropMaxTransIDSequenceNumber = "MaxTransIDSequenceNumber";
 static const char* const PropMaxTransSeedTimeoutMinutes = "MaxTransSeedTimeoutMinutes";
+static const char* const PropTransactionSeedType = "TransactionSeedType";
+static const char* const PropAlternativeTransactionSeedType = "AlternativeTransactionSeedType";
+static const char* const DefaultTransactionSeedType =  "-";
+static const char* const DefaultAlternativeTransactionSeedType =  "-X"; //local
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/logging/loggingagent/espserverloggingagent/loggingagent.cpp]: This is a default value. If someone does not want the -, it is configurable. I do not want to change the default, which may break existing users.

File: esp/logging/loggingagent/espserverloggingagent/loggingagent.hpp [hunk 3]
```diff
@@ -27,136 +27,18 @@
...
-        id.clear();
-
-        if (transIDFields)
-        {
-            add(transIDFields, sTransactionDateTime, id);
-            add(transIDFields, sTransactionMethod, id);
-            add(transIDFields, sTransactionIdentifier, id);
-        }
-        if (localSeed)
-            id.append(seed.get()).append("-X").append(++seq);
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/logging/loggingagent/espserverloggingagent/loggingagent.hpp]: This line has been removed.

File: esp/logging/loggingagent/espserverloggingagent/loggingagent.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaojx) [reply esp/logging/loggingagent/espserverloggingagent/loggingagent.cpp]: OK. Make sense

**Pull request merged**